src/core/ircprotocol.cpp

```diff
@@ -306,8 +306,25 @@ void IrcProtocolPrivate::handleCapabilityMessage(IrcCapabilityMessage* msg)
             }
             const QStringList params = msg->parameters();
             if (params.value(params.count() - 1) != QLatin1String("*")) {
-                if (!connection->saslMechanism().isEmpty() && availableCaps.contains(QLatin1String("sasl")))
-                    requestedCaps += QLatin1String("sasl");
+                if (!connection->saslMechanism().isEmpty()) {
+                    foreach (const QString& cap, availableCaps) {
+                        QStringList capParts = cap.split('=');
+                        if (capParts.length() == 2) {
+                            QString capName = capParts[0];
+                            if (capName.compare(QLatin1String("sasl"), Qt::CaseInsensitive) == 0) {
+                                // The server has advertised supporting SASL with a list of supported SASL Mechanisms, ensure our supported SASL Mechanism is part of that list before accepting the SASL capability
+                                QStringList serverSaslMethods = capParts[1].split(',');
+                                if (serverSaslMethods.contains(connection->saslMechanism())) {
+                                    requestedCaps += QLatin1String("sasl");
+                                }
+                            }
+                        } else {
+                            if (cap.compare(QLatin1String("sasl"), Qt::CaseInsensitive) == 0) {
+                                requestedCaps += QLatin1String("sasl");
+                            }
+                        }
```

Comment:
This could be merged to an `else if` to save an indentation layer :thinking: Or am I missing anything?
```suggestion
                        } else if (cap.compare(QLatin1String("sasl"), Qt::CaseInsensitive) == 0) {
                            requestedCaps += QLatin1String("sasl");
                        }
```